src/main/scala/io/iohk/ethereum/blockchain/sync/BlacklistSupport.scala

```diff
@@ -13,28 +12,46 @@ trait BlacklistSupport {
 
   def scheduler: Scheduler
 
-  var blacklistedPeers: Seq[(PeerId, Cancellable)] = Nil
+  private val maxSize = 1000
+
+  val blacklistedPeers = mutable.LinkedHashMap.empty[BlackListId, Cancellable]
 
-  def blacklist(peerId: PeerId, duration: FiniteDuration, reason: String): Unit = {
+  def blacklist(peerId: BlackListId, duration: FiniteDuration, reason: String): Unit = {
+
+    if (blacklistedPeers.size >= maxSize) {
+      removeOldestPeer()
+    }
     undoBlacklist(peerId)
     log.debug(s"Blacklisting peer ($peerId), $reason")
     val unblacklistCancellable = scheduler.scheduleOnce(duration, self, UnblacklistPeer(peerId))
-    blacklistedPeers :+= (peerId, unblacklistCancellable)
+    blacklistedPeers.put(peerId, unblacklistCancellable)
   }
 
-  def undoBlacklist(peerId: PeerId): Unit = {
-    blacklistedPeers.find(_._1 == peerId).foreach(_._2.cancel())
-    blacklistedPeers = blacklistedPeers.filterNot(_._1 == peerId)
+  def undoBlacklist(peerId: BlackListId): Unit = {
+    val peer = blacklistedPeers.get(peerId)
+    peer.foreach(_.cancel())
+    blacklistedPeers.remove(peerId)
   }
 
-  def isBlacklisted(peerId: PeerId): Boolean =
+  def isBlacklisted(peerId: BlackListId): Boolean =
     blacklistedPeers.exists(_._1 == peerId)
 
   def handleBlacklistMessages: Receive = {
     case UnblacklistPeer(ref) => undoBlacklist(ref)
   }
+
+  private def removeOldestPeer(): Unit = {
+    val oldestPeer = blacklistedPeers.head
+    oldestPeer._2.cancel()
+    blacklistedPeers.remove(oldestPeer._1)
+  }
 }
 
 object BlacklistSupport {
-  private case class UnblacklistPeer(peerId: PeerId)
+
+  abstract class BlackListId {
```

Comment:
tbh I was also a little torn apart here but ultimately decided to go with this abstract way because: 
1. It buys a little type safety
2. It keeps changes in the codebase to a minimum (in other case, we would need to call `peer.id.value` in all other `blacklist` calls)
3. taking off `extends AnyVal` from `PeerId` means there will some additional garbage associated with this change but I have never observed any bottlenecks near `PeerId` handling

Does this sound convincing?

Reply:
it does